### PR TITLE
Fix thread local.

### PR DESCRIPTION
**Pull request by zheng-da**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by eric-haibin-lin** [comment]
Could you add some explanation/doc so others won't be making the same mistake?

File: src/operator/nn/deconvolution.cu [hunk 1]
```diff
@@ -40,9 +40,15 @@ static CuDNNDeconvolutionOp<DType> &GetCuDNNDeconvOp(const DeconvolutionParam& p
                                                      const std::vector<TShape>& in_shape,
                                                      const std::vector<TShape>& out_shape,
                                                      const RunContext& rctx) {
+#if DMLC_CXX11_THREAD_LOCAL
```
Comment (eric-haibin-lin) [inline src/operator/nn/deconvolution.cu]: What about `dmlc::ThreadLocalStore`?

File: src/operator/nn/mkldnn/mkldnn_act.cc [hunk 2]
```diff
@@ -134,7 +134,11 @@ class MKLDNNActForward {
 static MKLDNNActForward &GetActForward(const ActivationParam& param,
                                        const OpContext &ctx, const NDArray &in_data,
                                        const mkldnn::memory &in_mem) {
+#if DMLC_CXX11_THREAD_LOCAL
   static thread_local std::unordered_map<MKLDNNActSignature, MKLDNNActForward, OpHash> fwds;
+#else
+  static MX_THREAD_LOCAL std::unordered_map<MKLDNNActSignature, MKLDNNActForward, OpHash> fwds;
```
Comment (anirudh2290) [inline src/operator/nn/mkldnn/mkldnn_act.cc]: do we need to make this change here too: https://github.com/apache/incubator-mxnet/blob/35fb9ea1093bd0006a396b5e493e5949b66ca33e/src/c_api/c_api_profile.cc#L115 ?

File: src/operator/nn/deconvolution.cu (hunk 1 above)
Comment (zheng-da) [reply src/operator/nn/deconvolution.cu]: not entirely sure. but the way we are using right now is to use MX_THREAD_LOCAL and thread_local depending on the macro.

File: src/operator/nn/mkldnn/mkldnn_act.cc (hunk 2 above)
Comment (zheng-da) [reply src/operator/nn/mkldnn/mkldnn_act.cc]: i think so.

**Comment by piiswrong** [comment]
Change dmlc to define MX_THREAD_LOCAL as thread_local when using cxx11?

**Comment by szha** [comment]
+1, way to go to reduce verbosity.

**Comment by szha** [comment]
That said, we can first merge this in so that follow-up release can be unblocked. Simplification of thread local usage is orthogonal.

**Pull request merged**